presto-hive/src/main/java/io/prestosql/plugin/hive/MsckProcedure.java

```diff
@@ -0,0 +1,213 @@
+{
+    private static final MethodHandle MSCK = methodHandle(
+            MsckProcedure.class,
+            "msck",
+            ConnectorSession.class,
+            String.class,
+            String.class,
+            Boolean.class,
+            Boolean.class);
+
+    private final Supplier<TransactionalMetadata> hiveMetadataFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+
+    @Inject
+    public MsckProcedure(
+            Supplier<TransactionalMetadata> hiveMetadataFactory,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "msck",
+                ImmutableList.of(
+                        new Argument("schema_name", VARCHAR),
+                        new Argument("table_name", VARCHAR),
+                        new Argument("add", BOOLEAN),
+                        new Argument("drop", BOOLEAN)),
+                MSCK.bindTo(this));
+    }
+
+    public void msck(ConnectorSession session, String schemaName, String tableName, Boolean add, Boolean drop)
+    {
+        HdfsEnvironment.HdfsContext context = new HdfsEnvironment.HdfsContext(session, schemaName, tableName);
+        SemiTransactionalHiveMetastore metastore = ((HiveMetadata) hiveMetadataFactory.get()).getMetastore();
+
+        Table table = metastore.getTable(schemaName, tableName).orElse(null);
+        if (table == null) {
+            return;
+        }
+
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation);
+
+            List<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName).orElse(emptyList());
+            List<String> partitionsInFileSystem = listDirectory(fileSystem, fileSystem.getFileStatus(defaultLocation), table.getPartitionColumns(), table.getPartitionColumns().size()).stream()
+                    .map(fileStatus -> defaultLocation.toUri().relativize(fileStatus.getPath().toUri()).getPath())
+                    .collect(toList());
+
+            // partitions in metastore but not in file system
+            List<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            // partitions in file system but not in metastore
+            List<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+
+            if (add) {
+                addPartitions(metastore, session, table, defaultLocation, partitionsToAdd);
+            }
+            if (drop) {
+                dropPartitions(metastore, session, table, partitionsToDrop);
+            }
+            metastore.commit();
```

Comment:
Under what circumstances the `metastore.rollback` should be called?

Reply:
I don't think `rollback` is needed here. It's a one-shot transaction(add and drop partitions in one transaction), so it's either all or nothing.

Reply:
I would prefer to have in nonetheless, just in case the code is going to evolve. I don't feel strongly on this.

Reply:
Hm.. i think actually _is_ needed -- https://github.com/prestosql/presto/pull/223#discussion_r256738452